### PR TITLE
bpo-30070: Fixed leaks and crashes in errors handling in the parser module.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**

### CODE REVIEW

**Comment by louisom** [comment]
@serhiy-storchaka May I ask a question? How did you found this bug, is via valgrind or something else? thanks.

**Comment by serhiy-storchaka** [comment]
Just by reading a code. I searched unchecked usages of PyObject_Size()/PySequence_Size()/PyMapping_Size() for bpo-30061 and after finding few suspicious cases in `parsermodule.c` analyzed other code in that file.

File: Modules/parsermodule.c [hunk 1]
```diff
@@ -1014,31 +1040,33 @@ build_node_tree(PyObject *tuple)
...
     else {
         /*  The tuple is illegal -- if the number is neither TERMINAL nor
          *  NONTERMINAL, we can't use it.  Not sure the implementation
          *  allows this condition, but the API doesn't preclude it.
          */
-        PyObject *err = Py_BuildValue("os", tuple,
+        PyObject *err = Py_BuildValue("Os", tuple,
                                       "Illegal component tuple.");
         PyErr_SetObject(parser_error, err);
         Py_XDECREF(err);
```
Comment (louisom) [inline Modules/parsermodule.c:1072]: Should `tuple` need `Py_DECREF` too? like what you done at line 950

**Review by louisom — COMMENTED** [review]
A small question about ref

File: Modules/parsermodule.c (hunk 1 above)
Comment (serhiy-storchaka) [reply Modules/parsermodule.c:1072]: No. `tuple` is a borrowed reference here (it's a parameter of `build_node_tree()`).

Comment (louisom) [reply Modules/parsermodule.c:1072]: ah, I see. didn't notice this. Thanks for your explain!

**Pull request merged**